### ISSUE TITLE
Addition of New Template 

### ISSUE THREAD

**Issue by Mad-u1** [invoice-x/invoice2data#154]
I have added two new template, but not able to incorporate the changes with the existing one.
Could you please guide on this. I have checked the below link also,
https://github.com/invoice-x/invoice2data/blob/master/TUTORIAL.rst

but no luck. Could please add some steps here, so I can communicate from my python project.

Thanks

**Comment by m3nu** [comment]
When using the library in Python, you need to load the plugin folder. By default it will load the built-in plugins only. The relevant function is `read_templates`, which is passed to the extraction function later. See [here](https://github.com/invoice-x/invoice2data/blob/d9af83d75c02bed90038c21eb5c8aa2f2e0a1288/invoice2data/main.py#L83) for an example.

If the steps to set up the Python code aren't mentioned in the README or tutorial yet, I agree they should be added.

**Comment by chriswakare** [comment]
@Mad-u1 
Let me know if the following code snippet helps?
"my_template_folder_name" is the custom template folder 
`from invoice2data import extract_data
from invoice2data.extract.loader import read_templates

read_template = read_templates(folder="my_template_folder_name")
result = extract_data('inv001.pdf', read_template)
print(result)`

**Comment by Mad-u1** [comment]
yes thanks for the code, but when have called the custom yml file,getting the below error,
 'Missing required key in template {} {}. Found {}'.format(name, path, tpl['fields'].keys())

attaching pdf and yml files here. please guide on this. 


[test_1.zip](https://github.com/invoice-x/invoice2data/files/2259862/test_1.zip)

thanks for your quick response.

**Comment by chriswakare** [comment]
At a glance, following fields are a must in the yml file to read invoice details from your :

amount
date
invoice_number

You can see a list of templates at extract/templates/com/
Pick up one that is nearest to your template and modify

**Comment by Mad-u1** [comment]
thanks for your guidance.
yes, this pdf only has the PO number.
if I enter any dummy data from the pdf in amount,date and invoice number to satisfy condition, then can I extract the PO number there?

**Comment by chriswakare** [comment]
As I recall, date has to be a valid date as the base code checks for a valid date
I never tried out amount as a non-decimal datatype - but looking at the output of my sample code - the JSON output for amount is a number

**Comment by Mad-u1** [comment]
thanks again, I'm able to extract all other pdfs but no luck for the shared pdf.
input_module.to_text(invoicefile).decode('utf-8') returns blank!!
** I'm able to add new attribute there to hold a new string data also.

**Comment by Mad-u1** [comment]
yes it is not working for embedded images. Have tried with tesseract also, but it seems tesseracts works with *.jpg files. not with pdf files.

**Comment by m3nu** [comment]
> yes, this pdf only has the PO number.
> if I enter any dummy data from the pdf in amount,date and invoice number to satisfy condition, then can I extract the PO number there?

We recently added a setting to specify required fields. If you only want the PO number that should be possible. Maybe there is still a bug left when logging the found field. Will reconfirm that.

> yes it is not working for embedded images. Have tried with tesseract also, but it seems tesseracts works with *.jpg files. not with pdf files.

For images you need to use one of the OCR modules. There is tesseract and also a Google Cloud OCR module (needs their API key). The default `pdftotext` input module will not work.

**Comment by Mad-u1** [comment]
yes I have already changed the input module to tesseract/pdf miner, but still it is not working for the shared pdf.
we can't use google or any 3rd party API due to policies.

**Comment by m3nu** [comment]
I see. That might be a limitation of the current workflow. Tesseract needs an image as input and we do some enhancements to the image before passing it on. To make it work with an image embedded in a PDF, the image needs to be extracted or PDF converted first. The latter may be most useful.

Leaves this todo list:

- [ ] Make sure logging doesn't need any required fields.
- [ ] If PDF is passed to Tesseract, convert to image first.

**Comment by Mad-u1** [comment]
thanks for the support. will update on this very shortly. will create filter before calling the Tesseract.
One more issue, if the pdf(without embedding)'s content is vertically rotated, need to know it is already rotated in coding section. Otherwise it is returning unwanted results. Need to handle that scenario also. Anything is there to realize it is already rotated, any flag or anything, then we can handle that issue with any filter.
thanks for your support.

**Comment by Mad-u1** [comment]
have created on filter to extract images from pdf(using pdftoimage). But can any one tell me, how to realize that the image in pdf is rotated or not..

thanks

**Comment by chriswakare** [comment]
Found the following link on https://github.com/tesseract-ocr/tesseract/wiki/ImproveQuality
Have not tried it though.

"Detecting rotation and line spacing of image of page of text using Radon transform "
https://gist.github.com/endolith/334196bac1cac45a4893#

https://gist.github.com/endolith/334196bac1cac45a4893#

**Comment by m3nu** [comment]
> Found the following link on https://github.com/tesseract-ocr/tesseract/wiki/ImproveQuality

@chriswakare improving OCR input is not fully solved. Feel free to experiment and make a PR to improve the command used in `tesseract.py`. This uses imagemagick for now. This would also be the place to convert PDFs to images.

**Comment by m3nu** [comment]
PS: Directly inputting a PDF works now. Small change to the convert-command.

**Comment by Mad-u1** [comment]
thanks u so much for the support, had used one filter there.. to extract the image. thanks mate.

@chriswakare 
Am checking mentioned link https://gist.github.com/endolith/334196bac1cac45a4893#.

thanks
